crates/data-generator/src/blocks/output.rs

```diff
@@ -0,0 +1,628 @@
+
+    impl ToTokens for BlockFromU32<'_> {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            let mut match_arms = TokenStream::new();
+            for (id, state_inst) in &self.0.id_states {
+                let state_inst = state_inst;
+
+                // "4287 => Farmland,"
+                let stream = quote! { #id => #state_inst, };
+                stream.to_tokens(&mut match_arms);
+            }
+
+            tokens.extend(quote! {
+                impl TryFrom<u32> for Block {
+                    type Error = super::UnknownId;
+
+                    fn try_from(id: u32) -> Result<Self, Self::Error> {
+                        Ok(match id {
+                            #match_arms
+                            _ => Err(super::UnknownId(id))?,
+                        })
+                    }
+                }
+            });
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct BlockStateInst {
+        pub properties: Vec<BlockStateProperty>,
+    }
+
+    impl From<&State> for BlockStateInst {
+        fn from(block_state: &State) -> Self {
+            let properties = block_state
+                .properties
+                .iter()
+                .map(|block_prop| BlockStateProperty {
+                    field_name: block_prop.field.to_ident(),
+                    enum_name: block_prop.prop_enum.to_ident(),
+                    var_name: block_prop.variant.to_ident(),
+                })
+                .collect();
+
+            BlockStateInst { properties }
+        }
+    }
+
+    impl ToTokens for BlockStateInst {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            // "Ladder"
+            tokens.append("Block".to_ident());
+
+            if !self.properties.is_empty() {
+                let mut props = TokenStream::new();
+                // "facing: Facing::North, waterlogged: Waterlogged::True,"
+                self.properties
+                    .iter()
+                    .for_each(|prop| prop.to_tokens(&mut props));
+
+                // "LargeAmethystBud { facing: Facing::North, waterlogged: Waterlogged::True, }"
+                tokens.append(Group::new(Delimiter::Brace, props))
+            }
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct BlockStateProperty {
+        pub field_name: Ident, // instrument
+        pub enum_name: Ident,  // NoteBlock
+        pub var_name: Ident,   // IronXylophone
+    }
+
+    impl ToTokens for BlockStateProperty {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            let field_name = &self.field_name;
+            let enum_name = &self.enum_name;
+            let var_name = &self.var_name;
+
+            // "instrument: NoteBlock::IronXylophone,"
+            tokens.extend(quote!(#field_name: #enum_name::#var_name,))
+        }
+    }
+
+    impl ToTokens for Id {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            tokens.append(Literal::u32_unsuffixed(self.0))
+        }
+    }
```

Comment:
This looks out of place here. Moving the implementation elsewhere won't help a whole lot; doing `Literal::u32_unsuffixed` at every call site isn't any better either.

Reply:
`u32` implements `ToTokens` too so maybe use that

Reply:
Unlike us, they make suffixed literals, that is, with `u32` suffix at the end.

Reply:
And so what? That *is* our protocol ID type, no?